// father/dist/builder/bundless/loaders/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.addLoader = void 0;
const utils_1 = require("@umijs/utils");
const fs_1 = __importDefault(require("fs"));
const loader_runner_1 = require("loader-runner");
const utils_2 = require("../../../utils");
const utils_3 = require("../../utils");
const dts_1 = require("../dts");
const loaders = [];
/**
 * add loader
 * @param item  loader item
 */
function addLoader(item) {
    // only support simple test type currently, because the webpack condition is too complex
    // refer: https://github.com/webpack/webpack/blob/0f6c78cca174a73184fdc0d9c9c2bd376b48557c/lib/rules/RuleSetCompiler.js#L211
    if (!['string', 'function'].includes(typeof item.test) &&
        !(item.test instanceof RegExp)) {
        throw new Error(`Unsupported loader test in \`${item.id}\`, only string, function and regular expression are available.`);
    }
    loaders.push(item);
}
exports.addLoader = addLoader;
/**
 * loader module base on webpack loader-runner
 */
exports.default = async (fileAbsPath, opts) => {
    const cache = (0, utils_2.getCache)('bundless-loader');
    // format: {path:contenthash:config:pkgDeps}
    const cacheKey = [
        fileAbsPath,
        (0, utils_3.getContentHash)(fs_1.default.readFileSync(fileAbsPath, 'utf-8')),
        JSON.stringify(opts.config),
        // use for babel opts generator in src/builder/utils.ts
        JSON.stringify(Object.assign({}, opts.pkg.dependencies, opts.pkg.peerDependencies)),
    ].join(':');
    const cacheRet = await cache.get(cacheKey, '');
    // use cache first
    /* istanbul ignore if -- @preserve */
    if (cacheRet) {
        const tsconfig = /\.tsx?$/.test(fileAbsPath)
            ? (0, dts_1.getTsconfig)(opts.cwd)
            : undefined;
        return Promise.resolve({
            ...cacheRet,
            options: {
                ...cacheRet.options,
                // FIXME: shit code for avoid invalid declaration value when tsconfig changed
                declaration: (tsconfig === null || tsconfig === void 0 ? void 0 : tsconfig.options.declaration) &&
                    (tsconfig === null || tsconfig === void 0 ? void 0 : tsconfig.fileNames.includes((0, utils_1.winPath)(fileAbsPath))),
            },
        });
    }
    // get matched loader by test
    const matched = loaders.find((item) => {
        switch (typeof item.test) {
            case 'string':
                return fileAbsPath.startsWith(item.test);
            case 'function':
                return item.test(fileAbsPath);
            default:
                // assume it is RegExp instance
                return item.test.test(fileAbsPath);
        }
    });
    if (matched) {
        // run matched loader
        return new Promise((resolve, reject) => {
            let outputOpts = {};
            (0, loader_runner_1.runLoaders)({
                resource: fileAbsPath,
                loaders: [{ loader: matched.loader, options: matched.options }],
                context: {
                    cwd: opts.cwd,
                    config: opts.config,
                    pkg: opts.pkg,
                    itemDistAbsPath: opts.itemDistAbsPath,
                    setOutputOptions(opts) {
                        outputOpts = opts;
                    },
                },
                readResource: fs_1.default.readFile.bind(fs_1.default),
            }, (err, { result }) => {
                if (err) {
                    reject(err);
                }
                else if (result) {
                    // FIXME: handle buffer type?
                    const ret = {
                        content: result[0],
                        options: outputOpts,
                    };
                    // save cache then resolve
                    cache.set(cacheKey, ret).then(() => {
                        resolve(ret);
                    });
                }
                else {
                    resolve(void 0);
                }
            });
        });
    }
};
